### PR TITLE
Fix refreshing keys from stale actors on signature verification error

### PR DESCRIPTION
**Pull request by ClearlyClaire**
`keypair_refresh_key!` had a case where it would return an account instead of a keypair, and a case where it would return a possibly incorrect key. This PR fixes these two issues and adds missing tests, as there were none exercising `keypair_refresh_key!`.

### CODE REVIEW

File: app/controllers/concerns/signature_verification.rb [hunk 1]
```diff
@@ -124,11 +133,9 @@ def stoplight_wrapper
   end
 
   def keypair_refresh_key!(keypair)
-    # TODO: this currently only is concerned with refreshing the actor and returning the legacy key, this needs to be reworked
     return if keypair.actor.local? || !keypair.actor.activitypub?
-    return keypair.actor.refresh! if keypair.actor.respond_to?(:refresh!) && keypair.actor.possibly_stale?
 
-    Keypair.from_legacy_account(ActivityPub::FetchRemoteActorService.new.call(keypair.actor.uri, only_key: true, suppress_errors: false))
+    ActivityPub::FetchRemoteKeyService.new.call(keypair.uri, suppress_errors: false, only_key: !keypair.actor.possibly_stale?)
```
Comment (oneiros) [inline app/controllers/concerns/signature_verification.rb]: I have a hard time concentrating on anything today, so I am probably not seeing the forest for trees. But could it be that neither this method here, nor `ActivityPub::FetchRemoteKeyService` persist the key that is being refreshed/fetched?

Comment (ClearlyClaire) [reply app/controllers/concerns/signature_verification.rb]: You are totally right, I thought this was handled by `ActivityPub::FetchRemoteActorService.new.call`, but this call is conditional. I'll split the first change out to a new PR and continue working on this.

Comment (ClearlyClaire) [reply app/controllers/concerns/signature_verification.rb]: Alright, this should be fixed now, sorry about that!

File: app/controllers/concerns/signature_verification.rb [hunk 2]
```diff
@@ -133,11 +133,22 @@ def stoplight_wrapper
   end
 
   def keypair_refresh_key!(keypair)
-    # TODO: this currently only is concerned with refreshing the actor and returning the legacy key, this needs to be reworked
     return if keypair.actor.local? || !keypair.actor.activitypub?
-    return keypair.actor.refresh! if keypair.actor.respond_to?(:refresh!) && keypair.actor.possibly_stale?
 
-    Keypair.from_legacy_account(ActivityPub::FetchRemoteActorService.new.call(keypair.actor.uri, only_key: true, suppress_errors: false))
+    actor = if keypair.actor.respond_to?(:refresh!) && keypair.actor.possibly_stale?
```
Comment (oneiros) [inline app/controllers/concerns/signature_verification.rb]: Why would the account not respond to `:refresh!`?

File: app/controllers/concerns/signature_verification.rb (hunk 1 above)
Comment (oneiros) [reply app/controllers/concerns/signature_verification.rb]: Sorry, for persisting here, but I still do not see it :frowning_face: 

I am thinking about the case in `ActivityPub::FetchRemoteKeyService` where the account is already known and not stale, but the key is not. I think it will be fetched and correctly returned in that case, but I do not see where it is being persisted.

Comment (ClearlyClaire) [reply app/controllers/concerns/signature_verification.rb]: Well, `FetchRemoteKeyService` would indeed not unconditionally fetch and persist the key, but we're not using it at all in the `keypair_refresh_key!` path. `FetchRemoteKeyService` would be used when we don't know of a key with this identifier. If we already know a key with this identifier, and the key is stale, it would go through `keypair_refresh_key!` which would refresh the account itself, fetching and updating keys in the process.

File: app/controllers/concerns/signature_verification.rb (hunk 2 above)
Comment (ClearlyClaire) [reply app/controllers/concerns/signature_verification.rb]: Currently it will always respond to `:refresh!`, but the code was written in a more generic way before this change already. This was from a while ago where we envisioned that we could handle different types of actors that would have different classes. If I remember correctly, this was when we were considering groups, and had groups as a new kind of actor that would be distinct from `Account` yet still use HTTP Signatures to relay messages around and stuff.

Comment (oneiros) [reply app/controllers/concerns/signature_verification.rb]: That explains the `actor` term being used here instead of `account` :bulb: 

But I would argue that most of the code here also expects this to be an `Account` object, so this special case here feels very confusing.

If you think it is worth keeping, what would you think about changing it to `keypair.actor.is_a?(Account)`? Not a lot less confusing, but maybe communicates the intent a bit better?

Comment (ClearlyClaire) [reply app/controllers/concerns/signature_verification.rb]: Let's remove this for now, and reintroduce such things if we ever have multiple classes.

File: app/controllers/concerns/signature_verification.rb (hunk 1 above)
Comment (oneiros) [reply app/controllers/concerns/signature_verification.rb]: The thing I had been missing was that `keypair_refresh_key!` here no longer makes use of `FetchRemoteKeyService`. I was fooled by that code snippet above this conversation, even though it *is* clearly marked as "outdated" :see_no_evil: 

Sorry for the confusion.

**Review by oneiros — APPROVED** [review]
